exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.drill.exec.vector.NullableVarCharVector;
+import org.apache.drill.exec.vector.ValueVector;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.drill.exec.store.pcap.PcapFormatUtils.parseBytesToASCII;
+
+public class PcapngRecordReader extends AbstractRecordReader {
+  private static final Logger logger = LoggerFactory.getLogger(PcapngRecordReader.class);
+
+  // batch size should not exceed max allowed record count
+  private static final int BATCH_SIZE = 40_000;
+
+  private static final Map<PcapngTypes, TypeProtos.MinorType> TYPES;
+
+  static {
+    TYPES = ImmutableMap.<PcapngTypes, TypeProtos.MinorType>builder()
+        .put(PcapngTypes.STRING, TypeProtos.MinorType.VARCHAR)
+        .put(PcapngTypes.INTEGER, TypeProtos.MinorType.INT)
+        .put(PcapngTypes.LONG, TypeProtos.MinorType.BIGINT)
+        .put(PcapngTypes.TIMESTAMP, TypeProtos.MinorType.TIMESTAMP)
+        .build();
+  }
+
+  private final Path pathToFile;
+  private OutputMutator output;
+  private ImmutableList<ProjectedColumnInfo> projectedCols;
+  private FileSystem fs;
+  private FSDataInputStream in;
+  private List<SchemaPath> projectedColumns;
+
+  private Iterator<IEnhancedPacketBLock> it;
+
+
+  public PcapngRecordReader(final String pathToFile,
+                            final FileSystem fileSystem,
+                            final List<SchemaPath> projectedColumns) {
+    this.fs = fileSystem;
+    this.pathToFile = fs.makeQualified(new Path(pathToFile));
+    this.projectedColumns = projectedColumns;
+  }
+
+  @Override
+  public void setup(final OperatorContext context, final OutputMutator output) throws ExecutionSetupException {
+    try {
+      this.output = output;
+      this.in = fs.open(pathToFile);
+      Decoder decoder = new Decoder(in);
+      this.it = decoder.getPcapngSectionList().iterator();
+      this.projectedCols = getProjectedColsIfItNull();
+      setColumns(projectedColumns);
+    } catch (IOException io) {
+      throw UserException.dataReadError(io)
+          .addContext("File name:", pathToFile.toUri().getPath())
+          .build(logger);
+    }
+  }
+
+  @Override
+  public int next() {
+    return parsePcapngFilesAndPutItToTable();
+  }
+
+  @Override
+  public void close() throws Exception {
+    in.close();
+  }
+
+  private ImmutableList<ProjectedColumnInfo> getProjectedColsIfItNull() {
+    return projectedCols != null ? projectedCols : initCols(new Schema());
+  }
+
+  private ImmutableList<ProjectedColumnInfo> initCols(final Schema schema) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+
+    for (String columnName : schema.getColumns().keySet()) {
```

Comment:
This code supports 'star' query, but with column order that defined in the ```Schema``` class. As I understand from previous comments we need to arrange the columns based on the projected list. Because of that, I will change this behaviour.